### ISSUE TITLE
RFDETR is not working in deepstream with fp16

### ISSUE THREAD

**Issue by vishnurm20011995-cmyk** [roboflow/rf-detr#445]
### Search before asking

- [x] I have searched the RF-DETR issues and found no similar bug report.

### Bug

I am using deepstream-7.0. I have used RFDETRSmall and RFDERTSeg in deepstream and was able to get the output when using fp32 (network_mode = 0). But when I change it to fp16 (network_mode = 2), the results are not coming properly. Occasionally some detections are coming, but still is having a very low accuracy compared to the fp32 version.

I created the onnx using the export function of rfdetr

### Environment

RF-DETR: 1.3.0
OS : Ubunty 20.04
Python version : 3.10.12
PyTorch version : 2.7.1
CUDA version : 12.4
GPU : NVIDIA GeForce RTX 2080

### Minimal Reproducible Example

#include <cstring>
 #include <iostream>
 #include "nvdsinfer_custom_impl.h"
 #include <cassert>
 #include <cmath>
 #include <algorithm>
 #include <map>
 #include <cstring>

#define MIN(a,b) ((a) < (b) ? (a) : (b))
#define MAX(a,b) ((a) > (b) ? (a) : (b))
#define CLIP(a,min,max) (MAX(MIN(a, max), min))
#define DIVIDE_AND_ROUND_UP(a, b) ((a + b - 1) / b)

extern "C"
bool NvDsInferParseCustomRFDETR (std::vector<NvDsInferLayerInfo> const &outputLayersInfo,
                                NvDsInferNetworkInfo const &networkInfo,
                                NvDsInferParseDetectionParams const &detectionParams,
                                std::vector<NvDsInferObjectDetectionInfo> &objectList) {

    // Code from NvDsInferParseCustomTfSSD for layer finding
    auto layerFinder = [&outputLayersInfo](const std::string &name)
        -> const NvDsInferLayerInfo *{
        for (auto &layer : outputLayersInfo) {
            if (layer.dataType == FLOAT &&
              (layer.layerName && name == layer.layerName)) {
                return &layer;
            }
        }
        return nullptr;
    };

    const NvDsInferLayerInfo *boxLayer = layerFinder("dets"); // 1 x num_queries x 4
    const NvDsInferLayerInfo *classLayer = layerFinder("labels");  // 1 x num_queries x num_classes

    if (!boxLayer || !classLayer) {
        std::cerr << "ERROR: some layers missing or unsupported data types "
                  << "in output tensors" << std::endl;
        return false;
    }

    const int keep_top_k = 300;
    unsigned int numDetections = classLayer->inferDims.d[0];
    unsigned int numClasses = classLayer->inferDims.d[1];
    std::map<float, NvDsInferObjectDetectionInfo> ordered_objects;

    size_t numClassesConfigured = detectionParams.perClassPreclusterThreshold.size();
    if(numClassesConfigured != numClasses) {
        std::cerr << "ERROR: numClassesConfigured: " << numClassesConfigured << " != numClasses from output tensor: " << numClasses << std::endl;
        return false;
    }

    for (unsigned int idx = 0; idx < numDetections; idx += 1) {
        NvDsInferObjectDetectionInfo res;

        unsigned int class_layer_idx = idx * numClasses;

        res.classId = std::max_element(((float*)classLayer->buffer+class_layer_idx), ((float*)classLayer->buffer+class_layer_idx+numClasses)) - ((float*)classLayer->buffer+class_layer_idx);
        res.detectionConfidence = ((float*)classLayer->buffer)[class_layer_idx+res.classId];

        // If model does not have sigmoid layer, perform sigmoid calculation here
        res.detectionConfidence = 1.0/(1.0 + exp(-res.detectionConfidence));

        if(res.detectionConfidence < detectionParams.perClassPreclusterThreshold[res.classId]) {
            continue;
        }
        enum {cx, cy, w, h};
        float rectX1f, rectY1f, rectX2f, rectY2f;

        unsigned int box_layer_idx = idx * 4;

        rectX1f = (((float*)boxLayer->buffer)[box_layer_idx + cx] - (((float*)boxLayer->buffer)[box_layer_idx + w]/2)) * networkInfo.width;
        rectY1f = (((float*)boxLayer->buffer)[box_layer_idx + cy] - (((float*)boxLayer->buffer)[box_layer_idx + h]/2)) * networkInfo.height;
        rectX2f = rectX1f + ((float*)boxLayer->buffer)[box_layer_idx + w] * networkInfo.width;
        rectY2f = rectY1f + ((float*)boxLayer->buffer)[box_layer_idx + h] * networkInfo.height;

        rectX1f = CLIP(rectX1f, 0.0f, networkInfo.width - 1);
        rectX2f = CLIP(rectX2f, 0.0f, networkInfo.width - 1);
        rectY1f = CLIP(rectY1f, 0.0f, networkInfo.height - 1);
        rectY2f = CLIP(rectY2f, 0.0f, networkInfo.height - 1);

        res.left = rectX1f;
        res.top = rectY1f;
        res.width = rectX2f - rectX1f;
        res.height = rectY2f - rectY1f;

        ordered_objects[res.detectionConfidence] = res;
    }

    int jdx = 0;
    for (auto iter=ordered_objects.rbegin(); iter!=ordered_objects.rend() && jdx<keep_top_k; iter++, jdx++) {
        if (iter->second.classId != 0){
        objectList.emplace_back(iter->second);}
    }
    return true;
}

CHECK_CUSTOM_PARSE_FUNC_PROTOTYPE(NvDsInferParseCustomRFDETR);

### Are you willing to submit a PR?

- [ ] Yes, I'd like to help by submitting a PR!

**Comment by vishnurm20011995-cmyk** [comment]
I had tried this. There was no /downsample/ layer in the verbose when I followed their method.
Below is the configuration I used
[property]
gpu-id=0
net-scale-factor=0.0039215697906911373
model-color-format=0
onnx-file=inference_model.sim.onnx
model-engine-file=inference_model_fp16.engine
labelfile-path=labels.txt
batch-size=1
network-mode=2
num-detected-classes=7
interval=0
gie-unique-id=1
process-mode=1
network-type=0
cluster-mode=2
maintain-aspect-ratio=1
symmetric-padding=1
workspace-size=4096
parse-bbox-func-name=NvDsInferParseCustomRFDETR
custom-lib-path=nvdsinfer_customparser_rfdetr.so

[class-attrs-all]
pre-cluster-threshold=0.25
topk=300

Note that I tried this with a RFDETRMedium model trained on a custom dataset. The FP32 version of this is giving good results

**Comment by MattLiutt** [comment]
@vishnurm20011995-cmyk I'm curious how you replace RFDETR with their default model in deepstream? like if you customize `nvdsinfer_customparser_rfdetr.so` yourself? Thanks!

**Comment by whittenator** [comment]
I did a little test where I pulled down [this](https://github.com/roboflow/single_artifact_benchmarking/tree/main) project and refactored it to only use my local RF-DETR Small ONNX model. It seemed to work properly, for the FP32 TensorRT model I got this:
```shell
Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.684
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.944
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.776
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.627
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.808
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.906
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.720
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.774
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.780
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.743
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.867
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.927
 Average Recall     (AR) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.994
 Average Recall     (AR) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.872
```

For FP16:
```shell
Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.683
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.944
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.771
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.625
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.809
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.906
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.718
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.773
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.779
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.741
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.867
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.923
 Average Recall     (AR) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.992
 Average Recall     (AR) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.875
```

Running COCO eval on the RF-DETR Pytorch I get this:
```shell
Average Precision (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.684
Average Precision (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.943
Average Precision (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.775
Average Precision (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.627
Average Precision (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ]  = 0.807
Average Precision (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ]  = 0.905
```

Obviously TensorRT version must be the issue. I am assuming in 10.4+ something was fixed to properly support the RF-DETR layers .

**Comment by vishnurm20011995-cmyk** [comment]
@whittenator I saw a similar issue in reddit : https://www.reddit.com/r/computervision/comments/1mwmexq/rfdetr_producing_wildly_different_results_with/
It seems the tensorrt version was the issue.

**Comment by isaacrob-roboflow** [comment]
Can I mark this as done? Seems like you got it working now :)

**Comment by Borda** [comment]
Apologies for the delayed close here — and great debugging by the community! The root cause was confirmed: TensorRT ≥10.4 is required for correct FP16 inference, and proper normalization values need to be set in the DeepStream config. Since this is an external TensorRT/DeepStream configuration issue rather than an RF-DETR bug, we're closing this. Thanks @whittenator for the thorough investigation — this thread is a great reference for anyone deploying to DeepStream.
